design/pipeline-resources.md

```diff
@@ -0,0 +1,199 @@
+
+```yaml
+resources:          # types: pipelines | repositories | containers | packages
+  repositories:
+  - repository: string # identifier for the repository resource      
+    type: enum # type of the repository source like AzureRepos, GitHub etc. In future this can extend to other source types
+    connection: string # service connection to connect to the source, defaults to primary source connection
+    source: string  # source repository to fetch
+    refs: string  # ref name to use, defaults to 'refs/heads/master'
+    clean: boolean  # whether to fetch clean each time
+    fetchDepth: number  # the depth of commits to ask Git to fetch
+    lfs: boolean  # whether to download Git-LFS files
+    submodules: true | recursive  # set to 'true' for a single level of submodules or 'recursive' to get submodules of submodules
+    persistCredentials: boolean  # set to 'true' to leave the OAuth token in the Git config after the initial fetch
+```
+
+### Examples
+
+```yaml
+resources:         
+  repositories:
+  - repository: secondaryRepo      
+    type: GitHub
+    connection: myGitHubConnection
+    source: Microsoft/alphaworz
+```
+
+### `checkout` your repository
+
+Repos from the `repository` resources defined are automatically synced and made available for all the jobs in the pipeline. However, in any of the jobs, you can choose to override and sync only specific repository using `checkout` shortcut. 
+
+```yaml
+- checkout: string  # identifier for your repository; for primary repository use the keyword self.
+  clean: boolean  # whether to fetch clean each time
+  fetchDepth: number  # the depth of commits to ask Git to fetch
+  lfs: boolean  # whether to download Git-LFS files
+  submodules: true | recursive  # set to 'true' for a single level of submodules or 'recursive' to get submodules of submodules
+  persistCredentials: boolean  # set to 'true' to leave the OAuth token in the Git config after the initial fetch
+  root: string        # directory to checkout the repo
+```
+
+The repo is checked-out to `$PIPELINES_RESOURCESDIR/<pipeline-identifier>/` folder.
+
+### Example
+
+```yaml
+- checkout: secondaryRepo  
+  clean: false
+  fetchDepth: 5
+  lfs: true
+```
+
+You can sync your primary repository as `self`.
+```yaml
+- checkout: self  
+```
+
+Or to avoid syncing any of the repository resources use:
+
+```yaml
+- checkout: none
+```
+
+When you use `checkout` to sync a specific repository resource, all the other repositories are not synced. 
+
+`self` checkout directory: `$PIPELINES_SOURCESDIR`
+other repositories' checkout directory: `$PIPELINES_RESOURCESDIR/<repository-identifier>/`
+
+## Resources: `containers`
+
+If you need to consume a container image as part of your CI/CD pipeline, you can achieve it using `containers`. A container can be an Azure Container Registy or any external Docker registry.
+
+### Schema
+
+```yaml
+resources:          # types: pipelines | repositories | containers | packages
+  containers:
```

Comment:
How does this match with what @vtbassmatt already has in the [Containers section](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=vsts&tabs=schema#container)? Are we changing `endpoint` to `connection`?

Reply:
yeah, today service endpoints are renamed to service connections. And wherever possible we are using the word `connection` instead of `endpoint`. I think we should change this.

Reply:
Agreed by @vtbassmatt. For backward compatibility we will still support `endpoint` as a deprecated input. But going forward we will use `connection`.